### ISSUE TITLE
4D vectors and quaternions

### ISSUE THREAD

**Issue by bvssvni** [PistonDevelopers/dyon#130]
The current size of `Variable` takes 24 bytes. This is because of `RustObject` and `Result<Box<Variable>, Box<Error>>` takes 16 bytes each.

It is possible to reduce the size of `Variable` to 16 bytes, but there is little performance gain:

```
24 bytes:
test tests::bench_add    ... bench:  44,400,761 ns/iter (+/- 8,135,966)
test tests::bench_add_n  ... bench:  16,062,670 ns/iter (+/- 3,626,578)
test tests::bench_array  ... bench:  18,056,946 ns/iter (+/- 2,723,787)
test tests::bench_call   ... bench:  16,998,256 ns/iter (+/- 3,153,922)
test tests::bench_len    ... bench:   3,175,400 ns/iter (+/- 165,141)
test tests::bench_main   ... bench:   2,022,940 ns/iter (+/- 205,091)
test tests::bench_n_body ... bench: 212,196,636 ns/iter (+/- 12,960,952)
test tests::bench_object ... bench:  27,209,689 ns/iter (+/- 4,189,969)
test tests::bench_sum    ... bench:   6,710,868 ns/iter (+/- 561,631)

16 bytes:
test tests::bench_add    ... bench:  42,593,998 ns/iter (+/- 6,324,724)
test tests::bench_add_n  ... bench:  15,714,274 ns/iter (+/- 2,906,590)
test tests::bench_array  ... bench:  17,976,067 ns/iter (+/- 2,897,692)
test tests::bench_call   ... bench:  17,438,250 ns/iter (+/- 3,088,261)
test tests::bench_len    ... bench:   3,184,736 ns/iter (+/- 125,076)
test tests::bench_main   ... bench:   1,950,572 ns/iter (+/- 95,575)
test tests::bench_n_body ... bench: 202,664,316 ns/iter (+/- 9,448,252)
test tests::bench_object ... bench:  26,567,412 ns/iter (+/- 3,930,720)
test tests::bench_sum    ... bench:   6,574,368 ns/iter (+/- 404,440)
```

For the n_body benchmark, there is a 5% improvement by reducing to 16 bytes. This might become greater after doing other optimizations.

However, if we keep 24 bytes, we can pack a `[f32; 4]` in `Variable`. This gives some interesting opportunities that might improve performance for game development.

### Motivation

4D vectors can be used for physical space-time and homogeneous coordinates. It is very nice to use for 3D programming.

A quaternion is a superset of complex numbers, with 3 complex dimensions instead of 1. They are commonly used to describe 3D rotations.

A matrix can be represented as `[vec4; 4]` and perform faster transformations because each operation requires less type checking.

### Axis notation

Complex numbers are very useful in many algorithms, and in mathematics this is the most commonly used notation:

```
a + b i
```

For quaternions, you have `i, j, k`:

```
a + b i + c j + d k
```

The variables `a, b, c, d` must be scalar numbers. `a` is the scalar component, and `b, c, d` are the vector components.

We could use the above notation to represent both complex numbers and quaternions (as quaternions).

However, `i, j, k` are often used as counters and this might reduce readability. Instead, we could use `ei, ej, ek`, borrowing from [Clifford algebra](https://en.wikipedia.org/wiki/Clifford_algebra):

```
a + b ei + c ej + d ek
```

For 4D vectors, we could use `ex, ey, ez, ew`:

```
a ex + b ey + c ez + d ew
```

When using number literals, you can type without space between the number and the axis:

```
1ex + 2ey + 3ez + 4ew
```

Just like with scientific notation, except the `e` means an axis instead of exponent. If you want `5*10^-3` along the x axis, you can type:

```
5e-3ex
```

You can not use `ex` and `ei` in the same expression, to separate 4D vectors from quaternions.

#### Example:

A point at `(4, 3, 0, 0)`:

```rust
fn main() {
    a := 4ex + 3ey
    println(a.x)
}
```

Complex number `2 + 4ei + 0ej + 0ek`:

```rust
fn main() {
    a := 2 + 4ei
    println(a) // prints `2 + 4ei`
}
```

### Tuple notation

This returns 4D vector, setting the remaining to 0.

```rust
a := (1, 2) // (1, 2, 0, 0)
a := (1, 2, 3) // (1, 2, 3, 0)
a := (1, 2, 3, 4)
```

These must be numbers.

Dyon is dynamically typed, and therefore uses `[1, 2, 3, 4]` where Rust would use tuples.

#### Example:

Change from 4D vector to 4D point in homogeneous coordinates:

```rust
fn main() {
    a := (0, 1, 2)
    a.w = 1
}
```

### Swizzling

When using two or more characters after `.`, e.g. `a.yx` it returns a 4D vector with values in the specified order.

#### Example:

```rust
fn main() {
    a := (1, 2, 3)
    b := a.zxy
    println(b) // prints `(3, 1, 2)`
}
```

### Indexing

For quaternion:

```rust
fn main() {
    q := 0 + 1ei + 2ej + 3k
    for i 4 {
        println(q[i]) // prints 0, 1, 2, 3
    }
}
```

For 4D vector:

```rust
fn main() {
    v := 0ex + 1ey + 2ez + 3ew
    for i 4 {
        println(v[i]) // prints 0, 1, 2, 3
    }
}
```

# Algebra

Common mathematical operators are built directly into the syntax of Dyon.

### Addition

Quaternion addition:

```rust
c := (1 + 2ei + 3ej + 4ek) + (5 + 6ei + 7ej + 8ek)
println(c) // prints `6 + 8ei + 10ej + 12ek`
```

4D vector addition:

```rust
c := (1, 2, 3, 4) + (5, 6, 7, 8)
println(c) // prints `(6, 8, 10, 12)`
```

### Per-component multiplication

Uses `*` as for normal numbers.

Quaternions:

```rust
a := 1 + 2ei + 3ej + 4ek
b := 5 + 6ei + 7ej + 8ek
c := a * b
println(c) // prints `5 + 12ei + 21ej + 32ek`
```

4D vectors:

```rust
a := (1, 2, 3, 4)
b := (5, 6, 7, 8)
c := a * b
println(c) // prints `(5, 12, 21, 32)`
```

### Norm

The norm is the same for both quaternions and 4D vectors, using `|` on each side:

```rust
a := (1, 2, 3, 4)
length := |a|
println(length)
```

Dyon does not use `|` so this works with the meta syntax.

### Cross product

The cross product uses `×` or `x`.

Ignores the last component for 4D vectors:

```rust
a := (1, 2, 3)
b := (4, 5, 6)
c := a × b
println(c) // prints `(_, _, _)`, because the last component is 0
```

Ignores the scalar component for quaternions:

```rust
a := 1ei + 2ej + 3ek
b := 4ei + 5ej + 6ek
c := a × b
println(c) // printlns `_ei + _ej + _ek`, because the scalar component is 0
```

### Dot product

Use `·` or `*.` to get the dot product of two 4D vectors:

```rust
a := (1, 2, 3)
b := (4, 5, 6)
c := a · b
```

### Scalar multiplication

```rust
a := 2 * (3ei + 4ej + 5ek) // quaternion
b := 2 * (3, 4, 5) // 4D vector
```

### Division

Only allowed on both sides, using per component:

```
a := (1, 2, 3) / 4
```

### Unary `-`

This is per component and follows same rule as for numbers.

### Conjugate

Quaternions use `~` to flip the sign complex dimensions:

```rust
a := ~(1+2ei+3ej+4ek)
println(a) // prints `1-2ei-3ej-4ek`
```

### Division by quaternion

Computes the reciprocal of right side and then performs a scalar multiplication with the left side:

```rust
a := 1+2ei+3ej+4ek
b := 1 / a
println(a * b) prints `1`
```

To multiply a quaternion with the reciprocal, use parenthesis `(1/q)`:

```rust
c := a * (1/b)
```

### Example: Reciprocal of a quaternion

```rust
a := 1+2ei+3ej+4ek
b := ~a / |a|^2
println(a * b) // prints `1`
println(b == 1/a) // prints `true`
```

### Example: Flip y component

```rust
a := (1, 2, 3)
a = a - 2*a.y ey
println(a) // prints `(1, -2, 3)`
```

**Comment by bvssvni** [comment]
Here are some mathematical unicode alternatives: `𝒾𝒿𝓀𝓍𝓎𝓏𝓉𝑤`

**Comment by bvssvni** [comment]
I don't think the axis notation is a good idea now, considering it overlaps quite a bit with addition.

The tuple notation together with swizzle notation makes the order much clearer:

```
(1, yz v) == (1, y(v), z(v), 0)
```

The indexing part suggested here is hard to do without adding run-time overhead, so I added a `s(v, ind)` intrinsic.

Neither is the swizzle notation suggested here easy to do without run-time overhead.

A better swizzle notation is to only allow it when creating a vector or calling a function. This makes it easier to put in other arguments, and write mathematical functions with numbers to get `f64` precision.

The quaternion stuff seems like adding a lot more complexity. I want to keep the language small. Could just use `vec4` and the scalar/vector representation of quaternions.

Closing since there are no more ideas here to explore.
